### PR TITLE
Add some new features and a new recipe for the enhancement task

### PR DESCRIPTION
**Pull request by Emrys365**
This PR is related to https://github.com/espnet/espnet/issues/2200.

1. Now different types of losses are supported for mask based networks, e.g. loss on enhanced spectrum, loss on predicted masks, etc.
2. A new recipe `egs2/mini_an4/enh1/` is created and also added to `ci/test_integration.sh` for testing.

### CODE REVIEW

**Comment by Emrys365** [comment]
An additional commit from @LiChenda is needed for fixing some issues in enh.sh.

File: ci/test_integration.sh [hunk 1]
```diff
@@ -157,6 +157,18 @@ done
 rm -rf exp dump data
 cd "${cwd}" || exit 1
 
+# [ESPnet2] test enh recipe
```
Comment (kamo-naoyuki) [inline ci/test_integration.sh:160]: Thank you! Could you also add validation part for the config files?

https://github.com/espnet/espnet/blob/933c94870a2ed6b1748fb34dbd13406b61f65a75/ci/test_integration.sh#L164-L177

Comment (Emrys365) [reply ci/test_integration.sh:160]: Done.

**Comment by LiChenda** [comment]
> An additional commit from @LiChenda is needed for fixing some issues in enh.sh.

It's solved.

**Comment by Emrys365** [comment]
> LGTM.
> I’m just curious why you use MSE for the mask loss? I thought that it would be better to use CE (in the binary case), KL or whatever (information theoretic loss).

Yes, I think we should make it flexible to choose any loss function that is appropriate for the selected loss type and mask type. I'll work on this later.
Currently, we are mainly testing the phase-sensitive mask (PSM), and I think it's okay to use mse loss for this mask type (according to [this paper](https://ieeexplore.ieee.org/document/7178061)).

**Comment by Emrys365** [comment]
It seems torch 1.1.0 is installed on the CI machine, so torchaudio won't be installed in this case.
Could you check it? @kamo-naoyuki 
https://github.com/espnet/espnet/pull/2238/checks?check_run_id=928189635#step:6:1158

**Comment by kamo-naoyuki** [comment]
So, you need to skip it in ci if <1.1.0.

File: espnet2/enh/layers/dnn_wpe.py [hunk 2]
```diff
@@ -84,11 +84,31 @@ def forward(
...
+            masks (torch.Tensor): (B, T, C, F)
+            ilens (torch.Tensor): (B,)
+        """
+        if self.use_dnn_mask:
+            (mask,), ilens = self.mask_est(data.permute(0, 3, 2, 1).float(), ilens)
+            # (B, F, C, T) -> (B, T, C, F)
+            mask = mask.transpose(-1, -3)
+        else:
+            mask = None
+        return mask, ilens
```
Comment (sas91) [inline espnet2/enh/layers/dnn_wpe.py:114]: In this file, you seem to be using mask_estimator from espnet1. I think using sigmoid activation might not be best for dereverberation mask. I think torch.clamp would be better. Can we pass arguments to control this activation instead of sigmoid being the only option?

Comment (sas91) [reply espnet2/enh/layers/dnn_wpe.py:114]: Also, for dereverberation, doing a direct signal approximation (instead of masking approximation) is better when trained as a pipeline. Is it possible to add this option?

**Comment by Emrys365** [comment]
> So, you need to skip it in ci if <1.1.0.

OK, now I add the conditional block.

File: espnet2/enh/layers/dnn_wpe.py (hunk 2 above)
Comment (Emrys365) [reply espnet2/enh/layers/dnn_wpe.py:122]: Thank you for your suggestions!

> Can we pass arguments to control this activation instead of sigmoid being the only option?

Yes. I have added two additional arguments to allow choosing the nonlinear type for both wpe and beamformer.

> Also, for dereverberation, doing a direct signal approximation (instead of masking approximation) is better when trained as a pipeline. Is it possible to add this option?

This is a bit complicated. Because the training loss is calculated in `espnet2/enh/espnet_model.py`. For now, it only gives an option to choose the loss type for all masks. If we want to choose the loss type for a specific module of a specific model, the code will be a bit messy.

A current workaround may be training the WPE and beamformer part (using different loss types) separately and combine them later.

**Comment by kamo-naoyuki** [comment]
ready to merge?

**Comment by Emrys365** [comment]
> ready to merge?

Yes, you can merge now. We will make a new PR later when the frontend-backend joint model is completed.

**Comment by kamo-naoyuki** [comment]
Thanks.

**Comment by kamo-naoyuki** [comment]
Hi. The design of `ESPnetEnhASRModel` is bad.
It is a combination of ESPnetASRModel and ESPnetEnhancementModel. In this case, delegate or inherit should be used according to common programming strategy.

- Delegate

```python
class ESPnetEnhASRModel(AbsESPnetModel):
    def __init__(
         self, 
         enh_model: ESPnetEnhancementModel,
         asr_model: ESPnetASRModel,
         ...
     ):
```

- Inherit

```python
class ESPnetEnhASRModel(ESPnetEnhancementModel, ESPnetASRModel):
```

Delegate is better.

**Comment by Emrys365** [comment]
Thank you for your suggestion! @shincling Could you try redesigning it?

**Comment by shincling** [comment]
Ok. I will do it.

**Comment by shincling** [comment]
> Thank you for your suggestion! @shincling Could you try redesigning it?

Ok. I will do it.
